doc: add a page on starting Flux

garlick opened the pull request:
This adds a  readthedocs page containing several ways to start a flux instance, with more explanation that we've given in our other higher level docs.

chu11 commented:
began reading this, I began thinking maybe this belongs more in flux-docs?

It might be a good "lead-in" document before some of the more specific tutorials, such as my ever pending https://github.com/flux-framework/flux-docs/pull/225

garlick commented:
Admittedly this section is useful stuff for flux framework generally but my goal (as noted in #5470) was to build up some flux-core specific documentation to complement the manual pages.

chu11 commented on doc/guide/start.rst:
```diff
@@ -0,0 +1,453 @@
+  portable to other systems that don't use Slurm as their primary resource
+  manager, as long as they can launch a Flux instance.
+
+- High throughput workloads don't have to bother the central Slurm controller
+  with every step, so they are less of a burden on the system and may run
+  faster.
```
maybe need to explain, "All slurm steps from all jobs in the system go through the central Slurm controller ...".

chu11 commented on doc/guide/start.rst:
```diff
@@ -0,0 +1,453 @@
+  <https://github.com/flux-framework/flux-sched>`_, Flux offers more
+  sophisticated step scheduling.
+
+- A Flux instance is highly configurable to meet the specific demands of
+  the workload.
```
.... "For example, Instead of the Fluxion scheduler, a scheduler specific to the workloads needs could be developed"??

chu11 commented on doc/guide/start.rst:
```diff
@@ -0,0 +1,453 @@
+  0: fluke6
+  1: fluke7
+
+The instance depth is zero since this is the system instance.  We also get a
+heads up about some unavailable nodes, and note that the instance owner is
+the ``flux`` user, but otherwise this looks like before.  Now, an interactive
+Flux allocation:
```
here and maybe this section, can show depth via `flux getattr instance-level`?

garlick replied:
`flux uptime` is already showing the depth in each example.  I wanted to stay focused on starting flux and not get that too diluted introducing other stuff.

chu11 commented on doc/guide/start.rst:
```diff
@@ -0,0 +1,453 @@
+
+- The local API socket is set to ``/run/flux/local``
+
+- The rank 0 broker is set become usable before all the other brokers have
+  joined, and to tolerate brokers joining and leaving as nodes go up and down.
```
english error in first sentence, not sure what was intent.  "The rank 0 broker must be usable before all other ..., so flux can tolerate ..."

garlick replied:
I meant "is set to become usable".  Will fix.

garlick commented on doc/guide/start.rst:
```diff
@@ -0,0 +1,453 @@
+  $ exit
+  $
+
+What just happened?  Hint: the process tree looks like this:
+
+.. code-block:: console
+
+  flux-start─┬─flux-broker-0───bash
+             └─flux-broker-1
+             └─flux-broker-2
```
This is just the process tree not a tree of jobs.

garlick commented on doc/guide/start.rst:
```diff
@@ -0,0 +1,453 @@
+   that it consists of three brokers, and that it has no parent instance, hence
+   a "depth" of zero.
+
+#. We run :man1:`flux-resource` *list* to show the instance's resource
+   inventory.  In this contrived test instance, we have three brokers running
+   on the same system, while on a real system we would usually have one.
+   Each broker assumes that it is discovering the resources of a unique
+   system, so the resource inventory contains contains three nodes and three
+   times the core count of the real node.
```
The point there is that one broker per node is the norm and the contrived setup is 3 brokers one node.  I'll see if I can reword that to be more clear.

garlick commented on doc/guide/start.rst:
```diff
@@ -0,0 +1,453 @@
+   The PMI server is only offered by :man1:`flux-start` when the
+   ``--test-size=N`` option is used.
+
+#. After the brokers are connected and have completed initialization, the
+   rank 0 broker spawns the initial program, an interactive shell that has its
+   environment set up so that Flux commands run from it will connect to the
+   new Flux instance.  The prompt right after the start command is from this
+   new shell.
```
Maybe :-) I'm on the fence on whether that's helpful here or not...

garlick commented:
I made a couple of changes to the last sections.

I dropped the LSF section since it appears that LSF users normally launch MPI programs with the `mpirun` for whatever MPI they are using, so a more appropriate section would be about how to start flux with the different `mpirun` variants.  I already have the one from mpich (hydra).  For now, I deleted the LSF section and we can leave the specialized `jsrun` instructions that apply only to LLNL and oak ridge to the coral document we already have.

I changed the "Starting with systemd" section to a more detailed and generalized "Starting with Static Configuration" section and mentioned the  operator and system instance in that section with external references.

chu11 reviewed (approved):
changes LGTM

garlick commented:
Thanks @chu11!
